Add CUDA 12 conda installation docs

bdice opened the pull request:
This PR adds documentation for CUDA 12 conda installation methods.

Closes #9045.

bdice commented:
cc: @jakirkham @pentschev @gmarkall for review. @gmarkall and anyone else with permission, feel free to push directly to this branch.

pentschev reviewed (commented):
LGTM, thanks @bdice !

jakirkham commented on docs/source/cuda/overview.rst:
```diff
@@ -60,9 +60,15 @@ Presently 11.2 is the minimum required toolkit version. An NVIDIA driver
+
+For CUDA 11, ``cudatoolkit`` is required::
+
+    $ conda install -c conda-forge cudatoolkit "cuda-version=11.*"
```
Maybe we can drop the `*` from this and similar lines

It is effectively implied. Also `*` sometimes confuses shells

```suggestion
    $ conda install -c conda-forge cudatoolkit "cuda-version=11"
```

jakirkham reviewed (commented):
Thanks Bradley! 🙏

This is a nice improvement 🙂

Added a few suggestions below

stuartarchibald commented:
Thanks for the patch @bdice and for the reviews @pentschev and @jakirkham. 

@gmarkall please could you take a look, Many thanks.

gmarkall commented on buildscripts/condarecipe.local/meta.yaml:
```diff
@@ -54,8 +54,8 @@ requirements:
-    # CUDA 10.2 or later is required for CUDA support
-    - cudatoolkit >=10.2
+    # CUDA 11.0 or later is required for CUDA support
+    - cudatoolkit >=11.0
```
Thanks for the suggestion - I think it should be 11.2, as Numba no longer works with the NVVM based on LLVM 3.4 that was in pre-11.2 toolkits
```suggestion
    - cuda-version >=11.2
    - cudatoolkit >=11.2
```

jakirkham replied:
Yeah was also wondering about this. We use [11.2 in conda-forge]( https://github.com/conda-forge/numba-feedstock/blob/aa6fdf29d00c437f7a112ed1650d56ef7ec247c0/recipe/meta.yaml#L56-L58 ) too

Was having trouble finding this in the docs though. Do we have the minimum written down somewhere?

gmarkall commented on docs/source/user/installing.rst:
```diff
@@ -45,12 +45,17 @@ Note that Numba, like Anaconda, only supports PPC in 64-bit little-endian mode.
+
+For CUDA 11, ``cudatoolkit`` is required::
 
+    $ conda install -c conda-forge cudatoolkit "cuda-version=11.*"
```
Similar to @jakirkham's suggestion above:
```suggestion
    $ conda install -c conda-forge cudatoolkit "cuda-version=11"
```

gmarkall commented on numba/cuda/cudadrv/nvvm.py:
```diff
@@ -416,10 +416,14 @@ def get_arch_option(major, minor):
 MISSING_LIBDEVICE_FILE_MSG = '''Missing libdevice file.
-Please ensure you have package cudatoolkit >= 11.0
-Install package by:
+Please ensure you have a CUDA Toolkit 11.0 or higher.
```
```suggestion
Please ensure you have a CUDA Toolkit 11.2 or higher.
```

gmarkall commented on numba/cuda/cudadrv/nvvm.py:
```diff
@@ -416,10 +416,14 @@ def get_arch_option(major, minor):
+
+For CUDA 11, ``cudatoolkit`` is required:
+
+    $ conda install -c conda-forge cudatoolkit "cuda-version=11.*"
```
As above:
```suggestion
    $ conda install -c conda-forge cudatoolkit "cuda-version=11"
```

jakirkham commented on docs/source/cuda/overview.rst:
```diff
@@ -60,9 +60,15 @@ Presently 11.2 is the minimum required toolkit version. An NVIDIA driver
+
+For CUDA 11, ``cudatoolkit`` is required::
+
+    $ conda install -c conda-forge cudatoolkit "cuda-version=11"
```
If we have an 11.2 minimum, would it make sense to use this?

```suggestion
    $ conda install -c conda-forge cudatoolkit "cuda-version>=11.2,<12"
```

jakirkham commented on docs/source/user/installing.rst:
```diff
@@ -45,12 +45,17 @@ Note that Numba, like Anaconda, only supports PPC in 64-bit little-endian mode.
+
+For CUDA 11, ``cudatoolkit`` is required::
 
+    $ conda install -c conda-forge cudatoolkit "cuda-version=11"
```
We may want to update this as well depending on how we handle the above comment

jakirkham commented:
Had some questions about incorporating the 11.2 minimum

Otherwise this looks good

bdice commented:
@gmarkall @jakirkham Sorry, I realized I had left off a crucial piece. I just updated the Minor Version Compatibility docs.

I also updated the minimum to 11.2 everywhere I could see and added a reference to CUDA 12.4 now that it's available. If we need to split this PR somehow, let me know.

jakirkham commented:
LGTM. Thanks Bradley! 🙏

gmarkall commented on docs/source/cuda/minor_version_compatibility.rst:
```diff
@@ -5,39 +5,61 @@ CUDA Minor Version Compatibility
+
+   pip install --extra-index-url https://pypi.nvidia.com pynvjitlink-cu12
+
+CUDA 12
```
```suggestion
CUDA 11
```

gmarkall reviewed (approved):
Looks good to me! Just waiting for CI prior to marking ready to merge.

bdice commented:
Looks like CI passed. @gmarkall This should be ready to merge.
